arxiv/vault/core.py

```diff
@@ -0,0 +1,217 @@
+"""Provides simple Vault API client."""
+
+from typing import Dict, Any, Optional, Union
+from pytz import UTC
+from datetime import datetime, timedelta
+from http import HTTPStatus
+import requests
+from json.decoder import JSONDecodeError
+
+import hvac
+from .hvac_extensions import Client
+from .adapter import HostnameLiberalAdapter
+
+import logging
+logger = logging.getLogger(__name__)
+logger.propagate = False
+
+
+class Secret:
+    """Represents a secret retrieved from Vault."""
+
+    def __init__(self, value: Any, issued: datetime, lease_id: str,
+                 lease_duration: int, renewable: bool) -> None:
+        """
+        Initialize a new secret.
+
+        Parameters
+        ----------
+        value : object
+            The value of the secret. May be a str value, or a struct of some
+            other kind.
+        issued : :class:`datetime`
+            The time that the secret was issued.
+        lease_id : str
+            Unique ID for the lease; can be used to renew the lease for the
+            secret.
+        lease_duration : int
+            Duration of the lease in seconds, starting at ``issued``.
+        renewable : bool
+            Whether or not the lease can be renewed.
+
+        """
+        self.value = value
+        self.issued = issued
+        self.lease_id = lease_id
+        self.lease_duration = lease_duration
+        self.renewable = renewable
+
+    @property
+    def expires(self) -> datetime:
+        """Get the datetime that the lease will expire."""
+        return self.issued + timedelta(seconds=self.lease_duration)
+
+    def is_expired(self, as_of: Optional[datetime] = None) -> bool:
+        """Check whether the token is expired (as of ``as_of``)."""
+        if as_of is None:
+            as_of = datetime.now(UTC)
+        return as_of >= self.expires
+
+
+class Token(Secret):
+    """An auth token secret."""
+
+
+class Vault:
+    """A simple Vault API client."""
+
+    kubernetes_mountpoint = 'kubernetes'
```

Comment:
Updated README with some notes on this:

> ### Mount point vs path vs key
>
> Vault has a whole bunch of pluggable functionality, and it is up to you to
enable the features (e.g. auth methods, secrets engines) that you want. When
you enable a  feature, you assign it to a **mount point**, a path on the Vault
API that will serve as the root path for the API of that feature. For example,
you could enable a [key-value secrets
engine](https://www.vaultproject.io/docs/secrets/kv/kv-v2.html) at the mount
point ``hopes-and-dreams/``; when [reading/writing
data](https://www.vaultproject.io/docs/secrets/kv/kv-v2.html#writing-reading-arbitrary-data),
for example, you would read and write to ``hopes-and-dreams/my-secret-key``.
>
> ```bash
> vault kv put hopes-and-dreams/my-secret-key my-value=s3cr3t
> ```
> 
> The **path** is the name of the secret relative to that mount point. So,
``my-secret-key`` in the example above is the path. Since kv secrets can
have more than one set of values, **key** refers to the key inside the secret
at path that contains the value of interest.